### PR TITLE
Fix OptionList and OptionSet to allow Iterable of Iterable

### PR DESCRIPTION
**Pull request by Jouramie**
## What is this fixing or adding?

image

## How was this tested?

Still have to figure out how item links work to test this, but yeah that should fix it.

## If this makes graphical changes, please attach screenshots.
N/A

### CODE REVIEW

**Comment by Jouramie** [comment]
@beauxq

**Comment by Jouramie** [comment]
Tested with two of this yaml
<details><summary>Test yaml</summary>
<p>

```yaml
Stardew Valley:
  item_links:
    - name: Pool party
      item_pool:
        - Summer
      replacement_item:
description: 'test'
game: Stardew Valley
name: jourardewvalley1
```
</p>
</details> 

<details><summary>Before</summary>
<p>

```
Traceback (most recent call last):
  File "C:\Users\Utilisateur\dev\Archipelago\Generate.py", line 407, in handle_option
    player_option = option.from_any(game_weights[option_key])
  File "C:\Users\Utilisateur\dev\Archipelago\Options.py", line 863, in from_any
    return cls.from_text(str(data))
  File "C:\Users\Utilisateur\dev\Archipelago\Options.py", line 856, in from_text
    return cls([option.strip() for option in text.split(",")])
  File "C:\Users\Utilisateur\dev\Archipelago\Options.py", line 86, in meta__init__
    self.value = self.schema.validate(self.value)
  File "C:\Users\Utilisateur\dev\Archipelago\venv\lib\site-packages\schema.py", line 366, in validate
    return type(data)(o.validate(d, **kwargs) for d in data)
  File "C:\Users\Utilisateur\dev\Archipelago\venv\lib\site-packages\schema.py", line 366, in <genexpr>
    return type(data)(o.validate(d, **kwargs) for d in data)
  File "C:\Users\Utilisateur\dev\Archipelago\venv\lib\site-packages\schema.py", line 167, in validate
    raise SchemaError(
schema.SchemaError: Or({'name': And(<class 'str'>, <built-in function len>), 'item_pool': [And(<class 'str'>, <built-in function len>)], Optional('exclude'): [And(<class 'str'>, <built-in function len>)], 'replacement_item': Or(And(<class 'str'>, <built-in function len>), None), Optional('local_items'): [And(<class 'str'>, <built-in function len>)], Optional('non_local_items'): [And(<class 'str'>, <built-in function len>)], Optional('link_replacement'): Or(None, <class 'bool'>)}) did not validate "[{'name': 'Pool party'"
"[{'name': 'Pool party'" should be instance of 'dict'
```
</p>
</details> 

<details><summary>After</summary>
<p>

```
Creating MultiWorld.
Creating Items.
Calculating Access Rules.
C:\Users\Utilisateur\dev\Archipelago\Utils.py:796: UserWarning: Getting options from multiworld is now deprecated. Please use `self.options.local_items` instead.
  warnings.warn(self.log_message)
C:\Users\Utilisateur\dev\Archipelago\Utils.py:796: UserWarning: Getting options from multiworld is now deprecated. Please use `self.options.non_local_items` instead.
  warnings.warn(self.log_message)
Running Item Plando.
Running Pre Main Fill.
Filling the multiworld with 808 items.
Balancing multiworld progression for 2 Players.
Beginning output...
Generating output files (2/2).
Calculating playthrough.
Creating final archive at C:\Users\Utilisateur\dev\Archipelago\output\AP_18222913313002705671.zip
Done. Enjoy. Total Time: 45.15276670000458
```

</p>
</details>

**Comment by beauxq** [comment]
https://github.com/ArchipelagoMW/Archipelago/pull/2874#discussion_r1513862720

**Comment by beauxq** [comment]
That name change is a little awkward.
`is_iterable_but_not_str` ? or `is_iterable_except_str` ?

File: Utils.py [hunk 1]
```diff
@@ -969,11 +969,8 @@ def __len__(self):
         return sum(len(iterable) for iterable in self.iterable)
 
 
-def is_iterable_of_str(obj: object) -> TypeGuard[typing.Iterable[str]]:
+def is_iterable_but_str(obj: object) -> TypeGuard[typing.Iterable[typing.Any]]:
     """ but not a `str` (because technically, `str` is `Iterable[str]`) """
```
Comment (beauxq) [inline Utils.py:973]:
```suggestion
    """ `str` is `Iterable`, but that's not what we want """
```
docstring could also be changed to match the change in what it does

**Comment by beauxq** [comment]
If `bytes` is also excluded, then I don't like the name as much, because it only brings attention to `str`.
maybe `is_non_primitive_iterable` ?

**Comment by Jouramie** [comment]
I'm not sure what exactly can be considered a primitive in python. In Java for instance it's pretty clear, primitives are not objects. But in python everything are objects, so...

**Review by alwaysintreble — APPROVED** [review]
change seems reasonable and tested that item links actually work now. didn't test any of the other affected options

**Review by agilbert1412 — APPROVED** [review]

**Comment by beauxq** [comment]
> I'm not sure what exactly can be considered a primitive in python. In Java for instance it's pretty clear, primitives are not objects. But in python everything are objects, so...

Another alternative is to reconsider whether `bytes` should be excluded.
I deal with `bytes` a lot and I haven't run into the same issues of finding it iterable, like I have with `str`.
The issue with `str` is special because because it's more difficult for type checkers to detect problems when iterating through a type gives you the same type.

Iterating through `str` gives you `str`.
- This can lead to confusing bugs, and confusing error messages.

Iterating through `bytes` doesn't give `bytes`, it gives `int`.
- This usually gives clear error messages if it's not what you want.

I think `bytes` should not be excluded from this function.

**Comment by Jouramie** [comment]
@zig-for I added bytes based on your input. Since iterating through `bytes` give you `int`, what do you think?

**Comment by black-sliver** [comment]
I'm with beauxq on this one. Even if there was a reason to specifically exclude bytes, then the function name is misleading.

**Review by agilbert1412 — APPROVED** [review]

**Comment by Jouramie** [comment]
I don't think there is a reason to exclude `bytes` right now. The current usage only requires excluding `str`.

**Review by black-sliver — APPROVED** [review]
Seems to fix the problem at hand and mypy is fine with the changed lines.

**Pull request merged**